### PR TITLE
support qwen3.5(vit) inference in turbomind backend

### PR DESCRIPTION
**Pull request by irexyc**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily receiving feedbacks. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by irexyc** [comment]
@copilot resolve the merge conflicts in this pull request

File: lmdeploy/archs.py [hunk 1]
```diff
@@ -133,10 +130,15 @@ def check_vl_llm(backend: str, config: dict) -> bool:
...
 
-def get_task(backend: str, model_path: str, trust_remote_code: bool = False):
+def get_task(backend: str,
+             model_path: str,
+             trust_remote_code: bool = False,
+             backend_config: PytorchEngineConfig | TurbomindEngineConfig | None = None):
     """Get pipeline type and pipeline class from model config."""
     from lmdeploy.serve.core import AsyncEngine
 
+    if backend_config and getattr(backend_config, 'disable_vision_encoder', False):
```
Comment (lvhan028) [inline lmdeploy/archs.py]: No "getattr"
```
not backend_config.disable_vision_encoder
```

File: lmdeploy/turbomind/builders/_base.py [hunk 2]
```diff
@@ -410,3 +410,33 @@ def _commit_tensor(self, name: str, tensor: torch.Tensor,
...
+    def __init__(self, config, ctx, *, root_handles):
+        super().__init__(config, ctx)
+        self._root_handles = root_handles
+
+    def build(self) -> BuiltModule:
+        built = super().build()
+        for i, (root, vis) in enumerate(zip(self._root_handles, built.handles)):
+            with self._ctx.devices[i]:
+                root.add_child_raw('visual_model', vis)
+        return built
```
Comment (lvhan028) [inline lmdeploy/turbomind/builders/_base.py]: Could we put this in a new file like `vision_model.py`? Also, for naming consistency, I prefer `VisionModelBuilder` over `VisualModelBuilder`.

File: lmdeploy/turbomind/builders/layer_norm.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+
+class LayerNormBuilder(Builder):
+    """Builder for a single LayerNorm weight module (weight + bias)."""
+
+    def set_weight(self, weight: torch.Tensor, bias: torch.Tensor | None = None):
+        """Commit the LayerNorm gamma (and optional beta) tensors."""
+        self._add_tensor('weight', weight)
+        if bias is not None:
+            self._add_tensor('bias', bias)
```
Comment (lvhan028) [inline lmdeploy/turbomind/builders/layer_norm.py]: We can put them into `norm.py`, as it is a type of norm.

File: lmdeploy/turbomind/models/qwen3_5.py [hunk 4]
```diff
@@ -1,30 +1,56 @@
...
 
+import math
 import re
-from typing import TYPE_CHECKING
+from typing import Any
 
 import _turbomind as _tm
+import torch
+from transformers.models.qwen3_5.configuration_qwen3_5 import Qwen3_5Config, Qwen3_5TextConfig
+from transformers.models.qwen3_5_moe.configuration_qwen3_5_moe import Qwen3_5MoeConfig, Qwen3_5MoeTextConfig
```
Comment (lvhan028) [inline lmdeploy/turbomind/models/qwen3_5.py]: Consider moving these imports to TYPE_CHECKING to avoid a runtime dependency on transformers.models.qwen3_5* module paths during import time (which can break across transformers versions).

File: lmdeploy/serve/processors/multimodal.py [hunk 5]
```diff
@@ -382,7 +382,18 @@ async def _get_multimodal_prompt_input(self,
...
-            results = await self.vl_encoder.preprocess(messages, mm_processor_kwargs)
+            if self.vl_encoder._uses_new_preprocess:
+                input_prompt = self.vl_encoder.model.get_input_prompt(messages=messages,
+                                                                      chat_template=chat_template,
+                                                                      sequence_start=sequence_start,
+                                                                      chat_template_kwargs=chat_template_kwargs)
+                results = await self.vl_encoder.preprocess(messages,
+                                                           input_prompt=input_prompt,
+                                                           mm_processor_kwargs=mm_processor_kwargs)
+                if getattr(self.vl_encoder.model, '_turbomind_native_vision', False):
```
Comment (lvhan028) [inline lmdeploy/serve/processors/multimodal.py]: Can we use `if self.vl_encoder.model._turbomind_native_vision`?

File: lmdeploy/vl/model/builder.py [hunk 6]
```diff
@@ -74,9 +74,11 @@ def load_vl_model(model_path: str,
                 logger.info(f'matching vision model: {name}')
                 model = module(**kwargs)
                 model.build_preprocessor(trust_remote_code=trust_remote_code)
-                # build the vision part of a VLM model when backend is
-                # turbomind, or load the whole VLM model when `with_llm==True`
-                if backend == 'turbomind' or with_llm:
+                # build the Python vision part for legacy TurboMind models,
+                # or load the whole VLM model when `with_llm==True`. Native
+                # TurboMind vision models run the visual encoder in C++.
+                native_tm_vision = backend == 'turbomind' and getattr(model, '_turbomind_native_vision', False)
```
Comment (lvhan028) [inline lmdeploy/vl/model/builder.py]: `model._turbomind_native_vision`

File: lmdeploy/turbomind/converter.py
Comment (lvhan028) [inline lmdeploy/turbomind/converter.py:150]: Now that `get_tm_config()` can construct multimodal models (not only text-only ones). But the docstring and local variable names still use `text_model`, which implies a narrower scope than the current behavior.

Could we rename these to `model` (e.g., return (model, model_path, data_type) and corresponding local names) to better reflect that this function now handles both text-only and multimodal model construction?

File: lmdeploy/turbomind/converter.py [hunk 7]
```diff
@@ -227,12 +231,26 @@ def get_tm_config(model_path,
     engine_config.mlp_tp_size = engine_config.mlp_tp_size or 1
 
     # 6. Build text model.
+    registered_name = get_registered_name(model_path, arch=arch)
+    model_cls = INPUT_MODELS.get(registered_name)
+
```
Comment (lvhan028) [inline lmdeploy/turbomind/converter.py]: Is there a specific reason to move get_registered_name() / INPUT_MODELS.get() above the cfg construction?

File: lmdeploy/turbomind/turbomind.py [hunk 8]
```diff
@@ -274,6 +275,20 @@ def _from_hf(self, model_path: str, engine_config: TurbomindEngineConfig,
...
+        if not multimodal:
+            return None
+        if self.engine_config.disable_vision_encoder:
+            logger.warning('Vision encoder has not been loaded, multimodal inputs will be ignored.')
+            return None
+
+        parser = getattr(self.source_model, 'to_turbomind_multimodal', None)
+        if parser is None:
+            raise ValueError(f'{type(self.source_model).__name__} does not support TurboMind multimodal inputs.')
+        return parser(multimodal)
```
Comment (lvhan028) [inline lmdeploy/turbomind/turbomind.py:289]: rename `parser` to `mm_input_converter` in this block for clearer intent. The callable here is not a generic parser; it converts frontend multimodal payloads into TurboMind-compatible multimodal inputs. `mm_input_converter` better matches the behavior and improves readability.

File: lmdeploy/turbomind/models/qwen3_5.py [hunk 9]
```diff
@@ -138,7 +164,8 @@ def reorder(x):
 
     def linear_attn(self, pfx):
         cfg = self._dn_cfg.clone()
-        builder = DeltaNetBuilder(cfg, self._ctx, tp=self._attn_tp)
+        builder = DeltaNetBuilder(cfg, self._ctx,
+                                  tp=self._attn_tp)
```
Comment (lvhan028) [inline lmdeploy/turbomind/models/qwen3_5.py]: No need to break it into two lines

File: lmdeploy/turbomind/models/qwen3_5.py [hunk 10]
```diff
@@ -233,25 +260,347 @@ def layers(self, pfx):
...
+        builder, keeping the visual sub-tree on its native dtype.
+        """
+        builder.config.data_type = self._resolver.data_type
+        return builder
+
+    def _build_visual_model(self, pfx):
+        cfg = self._make_visual_root_cfg()
+        root = self._restore_dtype(VisualModelBuilder(
+            cfg, self._ctx, root_handles=self._root_handles))
+        root.tp = self._model_tp
```
Comment (lvhan028) [inline lmdeploy/turbomind/models/qwen3_5.py]: Why not pass `self._model_tp` to `VisualModelBuilder` during initialization, just as `TextModelBuilder` does?

File: lmdeploy/turbomind/models/qwen3_5.py [hunk 11]
```diff
@@ -233,25 +260,347 @@ def layers(self, pfx):
...
+        cfg.temporal_patch_size = self._vis_temporal
+        cfg.num_position_embeddings = self._vis_pos_n
+        cfg.spatial_merge_size = self._vis_spatial_merge
+        cfg.norm_eps = self._vis_norm_eps
+        return cfg
+
+    def _patch_embed(self, pfx):
+        weight = pfx.pop('weight')
+        if weight.dim() >= 2:
+            weight = weight.reshape(weight.shape[0], -1).t().contiguous()
```
Comment (lvhan028) [inline lmdeploy/turbomind/models/qwen3_5.py]: I think we can remove "if weight.dim() >= 2" because it is always True, isn't it?

File: lmdeploy/turbomind/builders/__init__.py [hunk 12]
```diff
@@ -2,11 +2,12 @@
 """Builder sub-package — spec-driven module loading for TurboMind."""
 from __future__ import annotations
 
-from ._base import Builder, BuiltModule, SplitSide, _act_type_id, _torch_dtype_to_cpp
+from ._base import Builder, BuiltModule, SplitSide, VisualModelBuilder, _act_type_id, _torch_dtype_to_cpp
```
Comment (lzhangzz) [inline lmdeploy/turbomind/builders/__init__.py]: ALL visual -> vision

File: src/turbomind/models/qwen3_5vit/bias_gelu.cu [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+
+    PRAGMA_UNROLL
+    for (int i = 0; i < vec_size; ++i) {
+        x[i] = activation(x[i]);
+    }
+
+    Store(data + ti * stride + di, cast<T>(x));
+}
+
+}  // namespace
```
Comment (lzhangzz) [inline src/turbomind/models/qwen3_5vit/bias_gelu.cu:44]: Do not repeat the activation kernel

File: src/turbomind/models/language_model.cc [hunk 14]
```diff
@@ -406,9 +406,7 @@ void LanguageModel::Impl::Forward(int phase, TensorMap& env)
         Tensor input_embeds = LookupEmbedding(input_ids, symm_buf_);
         TM_DEBUG_TENSOR(input_embeds, "embeddings", 1);
 
-        auto& copy = *env.at("copy").data<BatchCopy*>()[0];
-        input_processor_->PatchEmbedding(phase, input_embeds, copy);
-        copy.Run();
```
Comment (lvhan028) [inline src/turbomind/models/language_model.cc:411]: Keep `copy.Run()` at outside

File: src/turbomind/python/CMakeLists.txt [hunk 15]
```diff
@@ -13,14 +13,7 @@ if(NOT pybind11_FOUND)
...
 pybind11_add_module(${PROJECT_NAME} bind.cpp)
-target_link_libraries(${PROJECT_NAME} PRIVATE xgrammar)
-# Force-link all objects from static libraries that contain self-registration
-# globals (module type registrars).  Without --whole-archive, the linker
-# strips registrar objects because nobody references them directly.
-target_link_libraries(${PROJECT_NAME} PRIVATE
-    "$<LINK_LIBRARY:WHOLE_ARCHIVE,models>"
-    "$<LINK_LIBRARY:WHOLE_ARCHIVE,turbomind>"
-)
+target_link_libraries(${PROJECT_NAME} PRIVATE turbomind xgrammar)
```
Comment (lvhan028) [inline src/turbomind/python/CMakeLists.txt:16]: check cmake version cc @lvhan028

File: src/turbomind/turbomind.cc [hunk 16]
```diff
@@ -276,9 +278,18 @@ void TurboMind::Impl::CreateEngine(int index)
     // create model
     LanguageModel model{param, ctx, *weights_[index]->text_model_ptr(), phases_};
 
+    // create optional visual model — the weight tree itself constructs
+    // its runtime peer (Factory Method); ``ModelRoot::visual_model`` is
+    // null for text-only checkpoints.
+    std::unique_ptr<VisualModel> visual_model;
+    if (auto* vw = weights_[index]->visual_model_ptr()) {
```
Comment (lvhan028) [inline src/turbomind/turbomind.cc]: auto vision_model = create_vision_model(...)

**Review by lvhan028 — APPROVED** [review]
